### PR TITLE
Support X509_is_equal functions

### PR DESCRIPTION
**Pull request by InfoHunter**
As discussed in #9091 , `X509_NAME_cmp` and related wrappers have ambiguous return values. Besides, more common use cases of the *cmp* series functions are checking if the two certificates or CRLs are the same instead of knowing which one is 'bigger' or 'smaller'.

So 3 new APIs are introduced in this commit: X509_is_equal(), X509_CRL_is_equal() and X509_NAME_is_equal(). The *equal* family only checks if the two compared entities are identical or not; unlike the `X509_cmp` family, who instead returns complex values indicating the *difference* of the compared entities.

Docs and test cases are added as well.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by InfoHunter** [comment]
Well, that could be better (and more simple probably). But in a previous comment of https://github.com/openssl/openssl/pull/9091#issuecomment-499461618 with @mspncp, it seems not possible at current stage to change to return value type. Besides, it seems it's still reasonable if we keep the 'ordering' return values 'as-is' in some use cases.

And this is why I reimplemented the `X509_NAME_is_equal` instead of wrapping it on a modified `X509_NAME_cmp`

**Comment by beldmit** [comment]
> Well, that could be better (and more simple probably). But in a previous comment of [#9091 (comment)](https://github.com/openssl/openssl/pull/9091#issuecomment-499461618) with @mspncp, it seems not possible at current stage to change to return value type. Besides, it seems it's still reasonable if we keep the 'ordering' return values 'as-is' in some use cases.

I do not suggest to change the return type of ```X509_NAME_cmp```, I suggest to make it more strict and wrap it in ```X509_NAME_is_equal```

**Comment by InfoHunter** [comment]
> I do not suggest to change the return type of X509_NAME_cmp, I suggest to make it more strict and wrap it in X509_NAME_is_equal

Then I think we need to make sure if we make it more strict we don't break the usages. For instance, someone compares A, B, C against D, he may get -10, -5, -3. Later he sorts (or whatever else operations) A, B, C based on the return values. In such case, making the return values to -1, 0, 1 will break the API.

Although I don't think the above example is very like to happen in practice, but I am still not very sure about making the X509_NAME_cmp return values stricter would be safe of not...

**Comment by InfoHunter** [comment]
Hmmm, but `X509_is_equal` can be change to wrap `X509_cmp`, I will fix that.

**Comment by mspncp** [comment]
>  For instance, someone compares A, B, C against D, he may get -10, -5, -3. Later he sorts (or whatever else operations) A, B, C based on the return values. In such case, making the return values to -1, 0, 1 will break the API.

I think your reasoning contains a little mistake. The sorting algorithms I had in mind (e.g. [qsort(3)](https://linux.die.net/man/3/qsort) don't rely on absolute values of the comparsion function's return value. In order to check whether A < B, A = B, or A > D, they wouldn't compare -10 against -5. Instead, they would call `compare(A, B)` and test the sign of its return value. 

So I don't see a problem in returning -1, 0, 1 if the sign of the results do not change. The problem in our case was that the return value was also 'misused' to return (negative) error codes.

**Comment by InfoHunter** [comment]
> Can X509_cmp return an error? If yes, we must return it too.

No, X509_cmp returns no error, only comparison result.

**Comment by InfoHunter** [comment]
> The sorting algorithms I had in mind (e.g. qsort(3) don't rely on absolute values of the comparsion function's return value.

Okay, I get the point. Even though, can we assume that it's safe enough to change the return value to `-1, 0, 1` combination in 3.0.0, given that there should be no (or rare) use cases of the absolute comparison values? Because the APIs' return values have indeed been changed anyway...

**Comment by mspncp** [comment]
Normally, it would not be important whether the comparison result is an arbitrary signed integer or just one of of the three values `{-1,0,+1}` as long as only the sign of the return value matters.

The real problem here is caused by the small number of comparison functions which return a negative error code (in most cases -2, but I'm not sure whether there are others). So -2 can be a comparison result, or an error code, depending on the function. It's this inconsistency which causes problems. You need to look up every single comparison function in the manual pages to find out whether you need error codes specially or not. And I'm not sure whether all these exceptional cases are documented properly in the manual pages. This was actually my point in https://github.com/openssl/openssl/pull/9091#issuecomment-499461618: We need to find out which error codes are possible and wich comparison functions return those.

Examples of functions returning -2 as error code are **X509_NAME_cmp**, **ASN1_TIME_cmp_time_t**, and **ASN1_UTCTIME_cmp_time_t**, and there are some comparison functions which forward their return values upwards. I've listed some examples below.

**X509_cmp** behaves properly, it only returns comparison results (not necessarily from `{-1,0,1}`)

https://github.com/openssl/openssl/blob/edc62356485257a37eb8775f25c5a19345b83a50/crypto/x509/x509_cmp.c#L134-L154


**X509_NAME_cmp** returns -2 to indicate an error:

https://github.com/openssl/openssl/blob/edc62356485257a37eb8775f25c5a19345b83a50/crypto/x509/x509_cmp.c#L156-L181


Some other comparison functions return the result of `X509_NAME_cmp()` directly in at least one of their code pathes:

https://github.com/openssl/openssl/blob/edc62356485257a37eb8775f25c5a19345b83a50/crypto/x509/x509_cmp.c#L18-L29

https://github.com/openssl/openssl/blob/edc62356485257a37eb8775f25c5a19345b83a50/crypto/x509/x509_cmp.c#L62-L75



**ASN1_STRING_cmp** returns arbitrary signed integers


https://github.com/openssl/openssl/blob/edc62356485257a37eb8775f25c5a19345b83a50/crypto/asn1/asn1_lib.c#L351-L364

There are other comparison function which return error codes.

**ASN1_TIME_cmp_time_t**

https://github.com/openssl/openssl/blob/edc62356485257a37eb8775f25c5a19345b83a50/crypto/asn1/a_time.c#L511-L530


**ASN1_UTCTIME_cmp_time_t**

https://github.com/openssl/openssl/blob/edc62356485257a37eb8775f25c5a19345b83a50/crypto/asn1/a_utctm.c#L72-L91



If you want to see more examples of comparison functions returning error codes, just run

```
git grep -nHW 'return -2'
```

(`-W` shows the function context of the grep hit!) and then search for '_cmp' in the pager (just type `/_cmp<ENTER>` to start the search and then repeatedly `n` to step throught the hits).

**Comment by mspncp** [comment]
Let's leave the problem of the error codes aside for a moment and ask the question: Do we really need a new set of comparison functions? Meanwhile, I have come to the conclusion that the benefit is not worth the effort.

Whether one prefers the semantics of `TYPE_cmp()` or `TYPE_is_equal()` is mostly a matter of taste. But note that the former functions yield more information; it is a one-liner to implement the latter given the first, but impossible to do it vice versa. Currently, OpenSSL has a lot of comparison functions of the form `TYPE_cmp()`. Introducing a new set of functions `TYPE_is_equal()` makes only sense if you do it for all of the existing comparison functions, not only to a few. Otherwise, inconsistency only increases.

The burdon to deal with error codes remains, even when switching to the new set: To make things consistent, callers of `TYPE_is_equal()` would have to check three types of return values 1:TRUE, 0:FALSE, <0:ERROR.

**Comment by InfoHunter** [comment]
> Let's leave the problem of the error codes aside for a moment and ask the question: Do we really need a new set of comparison functions? Meanwhile, I have come to the conclusion that the benefit is not worth the effort.

Well, the original motivation of introducing this set of `is_equal` API, is not to touch (or change) any bit of  the `_cmp` APIs and resolve the *error code* issues introduced in current `X509_NAME_cmp` and wrappers. So in this condition, it seems that adding new APIs would be the only way of resolving the problem.

But after some discussions, I find that if we consider changing the return values to `-1, 0, 1` combination is acceptable, then I think there is no need to add the new set of `_is_equal` APIs - because we can also resolve the *return value* problem in this way. So I agree with what @mspncp suggested.

**Comment by InfoHunter** [comment]
> Let's leave the problem of the error codes aside for a moment and ask the question: Do we really need a new set of comparison functions? Meanwhile, I have come to the conclusion that the benefit is not worth the effort.

Well, the original motivation of introducing this set of `is_equal` API, is not to touch (or change) any bit of  the `_cmp` APIs and resolve the *error code* issues introduced in current `X509_NAME_cmp` and wrappers. So in this condition, it seems that adding new APIs would be the only way of resolving the problem.

But after some discussions, I find that if we consider changing the return values to `-1, 0, 1` combination is acceptable, then I think there is no need to add the new set of `_is_equal` APIs - because we can also resolve the *return value* problem in this way. So I agree with what @mspncp suggested.

File: crypto/x509/x509_cmp.c [hunk 1]
```diff
@@ -74,6 +74,11 @@ int X509_CRL_cmp(const X509_CRL *a, const X509_CRL *b)
     return X509_NAME_cmp(a->crl.issuer, b->crl.issuer);
 }
 
+int X509_CRL_is_equal(const X509_CRL *a, const X509_CRL *b)
+{
+    return X509_NAME_is_equal(a->crl.issuer, b->crl.issuer);
+}
+
```
Comment (beldmit) [inline crypto/x509/x509_cmp.c:81]: I strongly disagree with the name of this function.

Comment (InfoHunter) [reply crypto/x509/x509_cmp.c:81]: So what is your suggestion then?

**Pull request closed without merging**